### PR TITLE
#1174 added balance voter

### PR DESCRIPTION
**Pull request by krzyk**
#1174 
* added balance voter with tests
* using it in elections with `1` weight (@g4s8 should it be higher?)

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1256](https://github.com/zerocracy/farm/issues/1256) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1256](https://github.com/zerocracy/farm/issues/1256) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/1771470178); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/zerocracy/pm/staff/votes/VsBalance.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        final double percent;
+        if (size == 0) {
+            percent = 0;
+        } else {
+            percent = (double) agenda.jobs(this.project).size() / size;
+        }
+        final int all = new LengthOf(
+            new Projects(this.farm, login).bootstrap().iterate()
+        ).intValue();
+        return (percent * all) / this.max.value();
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/pm/staff/votes/VsBalance.java]: @krzyk this formula seems wrong:

`votes = (((jobs in current project) * (total jobs)) * (number of projects)) / max(others' agendas)`

Whereas in https://github.com/zerocracy/farm/issues/1174#issue-333408214:

`vote = ((jobs in current project) / (total jobs)) * (total projects)`

Furthermore the original formula is wrong, if the intention is to boost the vote of projects where one has relatively less tasks compared with projects with more tasks. The unit test below actually shows this (see my other comment).

Cc @g4s8

File: src/test/java/com/zerocracy/pm/staff/votes/VsBalanceTest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        final Project first = new FkProject();
+        final Project second = new FkProject("SECOND123");
+        final FkFarm farm = new FkFarm();
+        final String user = "yegor256";
+        final Projects projects = new Projects(farm, user).bootstrap();
+        projects.add(first.pid());
+        new Agenda(farm, user).bootstrap()
+            .add(first, "none", "DEV");
+        MatcherAssert.assertThat(
+            new VsBalance(first, farm, new ListOf<>(user))
```
Comment (carlosmiranda) [inline src/test/java/com/zerocracy/pm/staff/votes/VsBalanceTest.java]: @krzyk the assertion here is exactly the opposite of the intention. In all of the user's projects:

* "first" project has one task
* "second" project has no tasks

The assertion actually made is that `first` will be ranked greater than `second`, which is the total opposite of what is intended. The formula is wrong (see my other comment)

File: src/main/java/com/zerocracy/pm/staff/votes/VsBalance.java (hunk 1 above)
Comment (krzyk) [reply src/main/java/com/zerocracy/pm/staff/votes/VsBalance.java]: @carlosmiranda OK, so the weight actually changes the vote result (I suggested myself by VsBanned voter). So here we need 1.0 - result, right?

Comment (krzyk) [reply src/main/java/com/zerocracy/pm/staff/votes/VsBalance.java]: @carlosmiranda as for the equotion, I think it is correct:

percent = jobs in current project / total jobs: `percent = (double) agenda.jobs(this.project).size() / size;`

all = number of projects of current user

and then according to https://github.com/zerocracy/farm/issues/1174#issuecomment-400283622 we normalize it with max number of projects that are currently assigned to one of the performers.

so the result is:
```
vote = ((jobs in current project) / (total jobs)) * (total projects) / (max projects assigned to single performer)
```

**Comment by krzyk** [comment]
@carlosmiranda updated code, and see my comment regarding the formula

File: src/main/java/com/zerocracy/pm/staff/votes/VsBalance.java (hunk 1 above)
Comment (carlosmiranda) [reply src/main/java/com/zerocracy/pm/staff/votes/VsBalance.java]: @krzyk thanks for the explanation. As for `VsBanned` it is weighted by -100 so positive results will mean lower vote, so `1.0 - result` that we have now is correct.

File: src/main/groovy/com/zerocracy/stk/pm/staff/elections/elect_performer.groovy [hunk 3]
```diff
@@ -100,10 +101,11 @@ def exec(Project project, XML xml) {
         (new VsSafe(new VsNoRoom(pmo)))                                           : role == 'REV' ? 0 : -100,
         (new VsSafe(new VsBanned(project, job)))                                  : -100,
         (new VsSafe(new VsVacation(pmo)))                                         : -100,
-        (new VsSafe(new VsWorkload(farm, logins)))                                 : 1,
-        (new VsSafe(new VsWorkload(farm, project, logins)))                        : 1,
+        (new VsSafe(new VsWorkload(farm, logins)))                                : 1,
+        (new VsSafe(new VsWorkload(farm, project, logins)))                       : 1,
         (new VsSafe(new VsSpeed(pmo, logins)))                                    : 3,
-        (new VsSafe(new VsRandom()))                                          : 1
+        (new VsSafe(new VsBalance(project, farm, logins)))                        : 1,
```
Comment (g4s8) [inline src/main/groovy/com/zerocracy/stk/pm/staff/elections/elect_performer.groovy]: @krzyk let's try to start with `3` here

File: src/test/java/com/zerocracy/pm/staff/votes/VsBalanceTest.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * @author Krzysztof Krason (Krzysztof.Krason@gmail.com)
+ * @version $Id$
+ * @since 0.23
+ * @checkstyle JavadocMethodCheck (500 lines)
+ */
+public final class VsBalanceTest {
+
+    @Test
+    public void returnsZeroForEmptyAgenda() throws IOException {
```
Comment (g4s8) [inline src/test/java/com/zerocracy/pm/staff/votes/VsBalanceTest.java]: @krzyk method name doesn't match expected result `Matchers.is(1.0)`, I'd rename it to `returnsHighestVoteForEmptyAgenda`

File: src/main/java/com/zerocracy/pm/staff/votes/VsBalance.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import java.io.IOException;
+import java.util.Collection;
+import org.cactoos.iterable.LengthOf;
+import org.cactoos.iterable.Mapped;
+import org.cactoos.scalar.IoCheckedScalar;
+import org.cactoos.scalar.MaxOf;
+import org.cactoos.scalar.SolidScalar;
+
+/**
+ * Project balance voter.
```
Comment (g4s8) [inline src/main/java/com/zerocracy/pm/staff/votes/VsBalance.java]: @krzyk please, add more details here, e.g. like in `VsSpeed`

**Comment by krzyk** [comment]
@g4s8 updated, please take a look

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1256](https://github.com/zerocracy/farm/issues/1256) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by carlosmiranda** [comment]
@ypshenychka confirmed

**Comment by ypshenychka** [comment]
@carlosmiranda Thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**